crates/sui-core/src/checkpoints/checkpoint_executor/mod.rs

```diff
@@ -198,182 +157,420 @@ impl CheckpointExecutor {
-
-                    self.process_executed_checkpoint(&epoch_store, &checkpoint, checkpoint_acc, checkpoint_data, &tx_digests, randomness_rounds).await;
-                    self.backpressure_manager.update_highest_executed_checkpoint(*checkpoint.sequence_number());
-                    highest_executed = Some(checkpoint.clone());
-
-                    // Estimate TPS every 10k transactions or 30 sec
-                    let elapsed = now_time.elapsed().as_millis();
-                    let current_transaction_num =  highest_executed.as_ref().map(|c| c.network_total_transactions).unwrap_or(0);
-                    if current_transaction_num - now_transaction_num > 10_000 || elapsed > 30_000 {
-                        let tps = (1000.0 * (current_transaction_num - now_transaction_num) as f64 / elapsed as f64) as i32;
-                        self.metrics.checkpoint_exec_sync_tps.set(tps as i64);
-                        now_time = Instant::now();
-                        now_transaction_num = current_transaction_num;
-                    }
-                     // we want to be inclusive of checkpoints in RunWithRange::Checkpoint type
-                    if run_with_range.map_or(false, |rwr| rwr.matches_checkpoint(checkpoint.sequence_number)) {
-                        info!(
-                            "RunWithRange condition satisfied after checkpoint sequence number {:?}",
-                            checkpoint.sequence_number
-                        );
-                        return StopReason::RunWithRangeCondition;
+        // check if we want to run this epoch based on RunWithRange condition value
+        // we want to be inclusive of the defined RunWithRangeEpoch::Epoch
+        // i.e Epoch(N) means we will execute epoch N and stop when reaching N+1
+        if run_with_range.map_or(false, |rwr| rwr.is_epoch_gt(self.epoch_store.epoch())) {
+            info!("RunWithRange condition satisfied at {:?}", run_with_range,);
+            return StopReason::RunWithRangeCondition;
+        };
+
+        self.metrics
+            .checkpoint_exec_epoch
+            .set(self.epoch_store.epoch() as i64);
+
+        let Some(next_to_schedule) = self.get_next_to_schedule() else {
+            return StopReason::EpochComplete;
+        };
+
+        let tps_estimator = Arc::new(Mutex::new(TPSEstimator::default()));
+
+        let this = Arc::new(self);
+
+        let is_final_checkpoint = stream_synced_checkpoints(
+            this.checkpoint_store.clone(),
+            next_to_schedule,
+            run_with_range.and_then(|rwr| rwr.into_checkpoint_bound()),
+        )
+        // concurrent step
+        .map(|checkpoint| {
+            checkpoint.report_checkpoint_age(&this.metrics.checkpoint_contents_age, &this.metrics.checkpoint_contents_age_ms);
+            this.backpressure_manager.update_highest_certified_checkpoint(*checkpoint.sequence_number());
```

Comment:
The comment here is sort of confusing, concurrent with what? Also calling this "is_final_checkpoint" seems misleading as this seems like a stream of some kind? in fact is this even used? it seems to be redefined below before its even used (if i'm reading this correctly)

Reply:
This is probably a bit hard to read, maybe I should factor things out so that the pipeline is visible in its entirety on one screen. I think it will become clearer then